### PR TITLE
Bump x/sys and opencontainers/runc to support Risc-V architecture

### PR DESCRIPTION
**Pull request by carlosedp**
**What type of PR is this?**

> /kind cleanup

**What this PR does / why we need it**:

Bump modules to allow building Kubernetes on Risc-V architecture.
Risc-V software support tracker on https://github.com/carlosedp/riscv-bringup

**Special notes for your reviewer**:

This PR depends on https://github.com/kubernetes/kubernetes/pull/82342 so it might need a rebase after that one get's merged.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go [hunk 1]
```diff
@@ -86,8 +82,23 @@ func newProp(name string, units interface{}) systemdDbus.Property {
...
+	fi, err := os.Lstat("/run/systemd/system")
+	if err != nil {
+		return false
+	}
+	return fi.IsDir()
+}
+
 func UseSystemd() bool {
-	if !systemdUtil.IsRunningSystemd() {
+	if !isRunningSystemd() {
```
Comment (BenTheElder) [inline vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go]: https://github.com/coreos/go-systemd/blob/fd7a80b32e1fc73e890fde45604ed5009dc817a3/util/util.go#L72

confirming this is the same method, just lifted into this package (no behavior change)

File: vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go [hunk 2]
```diff
@@ -208,10 +160,6 @@ func (m *Manager) Apply(pid int) error {
 
 	// if we create a slice, the parent is defined via a Wants=
 	if strings.HasSuffix(unitName, ".slice") {
-		// This was broken until systemd v229, but has been back-ported on RHEL environments >= 219
```
Comment (BenTheElder) [inline vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go]: possibly notable change https://github.com/opencontainers/runc/pull/2117

> Transient units (and transient slice units) have been available for quite a long time and RHEL 7 with systemd v219 (likely the oldest OS we care about at this point) supports that. A system running a systemd without these features is likely to break a lot of other stuff that runc/libcontainer care about.

File: vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go [hunk 3]
```diff
@@ -224,12 +172,7 @@ func (m *Manager) Apply(pid int) error {
 	}
 
 	// Check if we can delegate. This is only supported on systemd versions 218 and above.
-	if strings.HasSuffix(unitName, ".slice") {
-		if hasDelegateSlice {
-			// systemd 237 and above no longer allows delegation on a slice
-			properties = append(properties, newProp("Delegate", true))
-		}
```
Comment (BenTheElder) [inline vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go]: possibly notable change https://github.com/opencontainers/runc/pull/2117

> Regarding delegated slices, modern systemd doesn't allow it and runc/libcontainer run fine on it, so we might as well just stop requesting it on older versions of systemd which allowed it. (Those versions never really changed behavior significantly when that option was passed anyways.)

**Review by BenTheElder — COMMENTED** [review]
x/sys should be fine, runc has some possibly noteworthy changes

**Comment by carlosedp** [comment]
Any idea on why the failing tests @BenTheElder ?

**Comment by BenTheElder** [comment]
```
go: finding github.com/urfave/cli v1.21.0
go: finding golang.org/x/crypto v0.0.0-20190308221718-c2843e01d9a2
Skipping golang.org/x/... dependencies, pass --all to include

Use the given commands to remove pinned module versions that aren't actually used:
GO111MODULE=on go mod edit -dropreplace github.com/urfave/cli
```

**Comment by carlosedp** [comment]
Is this a CI hiccup or a real error... can't get form the logs.

**Comment by mattjmcnaughton** [comment]
@carlosedp k8s on RISC-V is some cool stuff!

**Comment by carlosedp** [comment]
Now it's GTG @BenTheElder :) Let's ping the approvers.

**Comment by BenTheElder** [comment]
I spoke to the upstream author re https://github.com/opencontainers/runc/pull/2117 and in theory no supported distros should have systemd old enough for this to be a problem.

Still might be release note worthy.
/lgtm
/approve

**Comment by carlosedp** [comment]
Folks, let me know whenever I should rebase this to be merged.
Thanks!

File: go.mod [hunk 4]
```diff
@@ -422,7 +422,7 @@ replace (
 	golang.org/x/net => golang.org/x/net v0.0.0-20190812203447-cdfb69ac37fc
 	golang.org/x/oauth2 => golang.org/x/oauth2 v0.0.0-20190402181905-9f3314589c9a
 	golang.org/x/sync => golang.org/x/sync v0.0.0-20181108010431-42b317875d0f
-	golang.org/x/sys => golang.org/x/sys v0.0.0-20190209173611-3b5209105503
```
[the remark this answers is not in the document]
Comment (liggitt) [reply go.mod]: go1.13 support landed in https://github.com/kubernetes/kubernetes/pull/82809

**Comment by liggitt** [comment]
Rebasing on master should shrink the size of this. I still would not expect to run ahead of our go release for golang.org/x/sys unless there was a strong reason to do so. Is there a discussion around prioritization of adding support for this architecture that I'm missing?

**Comment by carlosedp** [comment]
No priorization on the arch. I've been building the packages locally and applying changes. I'll demo K8s on Risc-V arch on the Risc-V Summit now in december so I came to check the status. 

K8s 1.17 will be built with 1.13 already?

One thing I noticed is that the `release-branch.go1.13` of `x/sys` misses a bugfix commit. I've pinged folks there if it can be applied.

**Comment by carlosedp** [comment]
@liggitt I've just checked the commits below and they have already bumped up the dependencies on `master`. I believe 1.18 will be cut from this, right?

https://github.com/kubernetes/kubernetes/commit/660b17d0aeda96af94defd4c5110d9fef523d52b
https://github.com/kubernetes/kubernetes/commit/708a917b7d76b704f706998442caf5404024740e

With this, I believe this PR can be closed.

**Comment by liggitt** [comment]
1.17 will be cut with go1.13
1.18 will be cut from what is in master

**Comment by carlosedp** [comment]
Hi @liggitt , I've just built K8s with latest master and it worked fine. All libraries have been bumped. I need to update `hack/lib/golang.sh` adding riscv64 to the list of architectures so `make all KUBE_BUILD_PLATFORMS=linux/riscv64` works fine. I'll submit another PR and close this, ok?

**Comment by liggitt** [comment]
I'm unsure of all the considerations for supporting a new architecture… does that mean we claim test/support for it?

**Comment by dims** [comment]
@liggitt we already support builds for ppc64le, s390x etc and we don't claim "support" for those arches. so this would be similar to those i think

**Comment by carlosedp** [comment]
Exactly, It's not claiming support or officially providing images right now but the ability to correctly build the packages using official pipeline and scripts.

**Comment by liggitt** [comment]
> Exactly, It's not claiming support or officially providing images right now but the ability to correctly build the packages using official pipeline and scripts.

Ok. I'll defer to the owners of those build/release scripts for approval on the follow up PR.

**Comment by BenTheElder** [comment]
> Hi @liggitt , I've just built K8s with latest master and it worked fine. All libraries have been bumped. I need to update hack/lib/golang.sh adding riscv64 to the list of architectures so make all KUBE_BUILD_PLATFORMS=linux/riscv64 works fine. I'll submit another PR and close this, ok?

👍, happy to review this

**Comment by carlosedp** [comment]
Latest master branch builds correctly on riscv64.

@BenTheElder I've opened #86011 to address the build script.

Closing this.

**Pull request closed without merging**